src/graph/sampler.cc

```diff
@@ -991,6 +991,118 @@ Subgraph NegEdgeSubgraph(int64_t num_tot_nodes, const Subgraph &pos_subg,
   return neg_subg;
 }
 
+Subgraph PBGNegEdgeSubgraph(int64_t num_tot_nodes, const Subgraph &pos_subg,
+                            const std::string &neg_mode,
+                            int neg_sample_size, bool is_multigraph,
+                            bool exclude_positive) {
+  unsigned int seed = randseed();
+  std::vector<IdArray> adj = pos_subg.graph->GetAdj(false, "coo");
+  IdArray coo = adj[0];
+  int64_t num_pos_edges = coo->shape[0] / 2;
+
+  int64_t chunk_size = neg_sample_size;
+  int64_t num_chunks = num_pos_edges / chunk_size;
+  int64_t last_chunk_size = num_pos_edges - num_chunks * chunk_size;
+  int64_t num_neg_edges = (chunk_size * 2 - 1) * chunk_size * 2 * num_chunks;
```

Comment:
should we use "num_chunks - 1" here? It seems you deal with last chunk differently.